File: defence_claw/defenseclaw/internal/observability/redaction/hash_v1.go
// Copyright 2026 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// SPDX-License-Identifier: Apache-2.0

// Package redaction implements value-safe observability redaction primitives.
package redaction

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"net/netip"
	"regexp"
	"strings"
	"unicode/utf8"

	"github.com/defenseclaw/defenseclaw/internal/observability"
	"golang.org/x/text/unicode/norm"
)

const (
	hashV1Domain  = "defenseclaw-redaction-hash-v1"
	hashV1KeySize = 32
)

// HashV1ErrorCode is a value-free reason why hash-v1 could not be applied.
type HashV1ErrorCode string

const (
	HashV1ErrorInvalidUTF8       HashV1ErrorCode = "invalid_utf8"
	HashV1ErrorInvalidKey        HashV1ErrorCode = "invalid_key"
	HashV1ErrorUnsupportedClass  HashV1ErrorCode = "unsupported_class"
	HashV1ErrorUnicodeRepertoire HashV1ErrorCode = "unicode_repertoire"
	HashV1ErrorNormalization     HashV1ErrorCode = "normalization_failed"
)

// HashV1Error deliberately contains no input value, key material, or key ID.
type HashV1Error struct {
	Code HashV1ErrorCode
}

func (e *HashV1Error) Error() string {
	return "redaction hash-v1 failed: " + string(e.Code)
}

// IsHashV1Error reports whether err has the requested safe hash-v1 code.
func IsHashV1Error(err error, code HashV1ErrorCode) bool {
	var typed *HashV1Error
	return errors.As(err, &typed) && typed.Code == code
}

func hashV1Error(code HashV1ErrorCode) error {
	return &HashV1Error{Code: code}
}

// HashV1 applies the version-1, cross-language keyed correlation transform.
// It returns only the non-reversible token; normalized input is never exposed.
func HashV1(value string, fieldClass observability.FieldClass, key []byte) (string, error) {
	digest, err := hashV1Digest(value, fieldClass, key)
	if err != nil {
		return "", err
	}
	keyID := hashV1KeyID(key)

	return fmt.Sprintf(
		"<hashed class=%s v=1 key=%s len=%d hmac=%s>",
		fieldClass,
		keyID,
		len(value),
		hex.EncodeToString(digest[:]),
	), nil
}

// hashV1Digest is the shared keyed transform behind both the formatted
// projection token and the compact finding-correlation fingerprint. Keeping
// the transform here guarantees identical version, normalization, domain, and
// field-class separation without exposing key material from Engine.
func hashV1Digest(
	value string,
	fieldClass observability.FieldClass,
	key []byte,
) ([sha256.Size]byte, error) {
	var digest [sha256.Size]byte
	if !utf8.ValidString(value) {
		return digest, hashV1Error(HashV1ErrorInvalidUTF8)
	}
	if len(key) != hashV1KeySize {
		return digest, hashV1Error(HashV1ErrorInvalidKey)
	}
	if !observability.IsFieldClass(fieldClass) {
		return digest, hashV1Error(HashV1ErrorUnsupportedClass)
	}

	normalized, err := normalizeHashV1Value(value, fieldClass)
	if err != nil {
		return digest, err
	}
	mac := hmac.New(sha256.New, key)
	_, _ = mac.Write([]byte(hashV1Domain))
	_, _ = mac.Write([]byte{0})
	_, _ = mac.Write([]byte(fieldClass))
	_, _ = mac.Write([]byte{0})
	_, _ = mac.Write([]byte(normalized))
	copy(digest[:], mac.Sum(nil))
	return digest, nil
}

func hashV1KeyID(key []byte) string {
	digest := sha256.Sum256(key)
	return hex.EncodeToString(digest[:6])
}

func normalizeHashV1Value(value string, fieldClass observability.FieldClass) (string, error) {
	if !utf8.ValidString(value) {
		return "", hashV1Error(HashV1ErrorInvalidUTF8)
	}
	if !observability.IsFieldClass(fieldClass) {
		return "", hashV1Error(HashV1ErrorUnsupportedClass)
	}
	if !isUnicode13Repertoire(value) {
		return "", hashV1Error(HashV1ErrorUnicodeRepertoire)
	}
	value = norm.NFC.String(value)
	if fieldClass != observability.FieldClassPath {
		return value, nil
	}
	if isWindowsDrivePath(value) {
		return normalizeLexicalPath(value), nil
	}
	if hierarchicalURIPattern.MatchString(value) {
		return normalizeAbsoluteURI(value)
	}
	return normalizeLexicalPath(value), nil
}

var hierarchicalURIPattern = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9+.-]*://`)

func isWindowsDrivePath(value string) bool {
	return len(value) >= 2 && isASCIILetter(value[0]) && value[1] == ':'
}

func normalizeLexicalPath(value string) string {
	value = strings.ReplaceAll(value, `\`, "/")

	prefix := ""
	absolute := false
	uncRoot := false
	rest := value
	switch {
	case len(rest) >= 2 && rest[0] == '/' && rest[1] == '/':
		prefix = "//"
		rest = strings.TrimLeft(rest[2:], "/")
		parts := nonemptyPathSegments(rest)
		if len(parts) >= 2 && isOrdinaryPathSegment(parts[0]) && isOrdinaryPathSegment(parts[1]) {
			prefix += parts[0] + "/" + parts[1]
			rest = strings.Join(parts[2:], "/")
			absolute = true
			uncRoot = true
		}
	case strings.HasPrefix(rest, "/"):
		prefix = "/"
		absolute = true
		rest = strings.TrimLeft(rest[1:], "/")
	case len(rest) >= 2 && isASCIILetter(rest[0]) && rest[1] == ':':
		prefix = strings.ToLower(rest[:1]) + ":"
		rest = rest[2:]
		if strings.HasPrefix(rest, "/") {
			prefix += "/"
			absolute = true
			rest = strings.TrimLeft(rest, "/")
		}
	}

	segments := make([]string, 0, strings.Count(rest, "/")+1)
	for _, segment := range strings.Split(rest, "/") {
		switch segment {
		case "", ".":
			continue
		case "..":
			if len(segments) > 0 && segments[len(segments)-1] != ".." {
				segments = segments[:len(segments)-1]
			} else if !absolute {
				segments = append(segments, segment)
			}
		default:
			segments = append(segments, segment)
		}
	}

	joined := strings.Join(segments, "/")
	if prefix == "" {
		return joined
	}
	if joined == "" {
		return prefix
	}
	if strings.HasSuffix(prefix, "/") {
		return prefix + joined
	}
	if uncRoot {
		return prefix + "/" + joined
	}
	return prefix + joined
}

func nonemptyPathSegments(value string) []string {
	parts := strings.Split(value, "/")
	result := make([]string, 0, len(parts))
	for _, part := range parts {
		if part != "" {
			result = append(result, part)
		}
	}
	return result
}

func isOrdinaryPathSegment(value string) bool {
	return value != "" && value != "." && value != ".."
}

func normalizeAbsoluteURI(value string) (string, error) {
	if !isASCIIURI(value) {
		return "", hashV1Error(HashV1ErrorNormalization)
	}
	colon := strings.IndexByte(value, ':')
	if colon <= 0 {
		return "", hashV1Error(HashV1ErrorNormalization)
	}
	scheme := strings.ToLower(value[:colon])
	rest := value[colon+1:]
	if !strings.HasPrefix(rest, "//") {
		return "", hashV1Error(HashV1ErrorNormalization)
	}

	// A fragment is syntactically validated even though it is discarded.
	if fragmentAt := strings.IndexByte(rest, '#'); fragmentAt >= 0 {
		if _, err := normalizeURIComponent(rest[fragmentAt+1:], uriQueryFragment); err != nil {
			return "", err
		}
		rest = rest[:fragmentAt]
	}

	query := ""
	hasQuery := false
	if queryAt := strings.IndexByte(rest, '?'); queryAt >= 0 {
		hasQuery = true
		var err error
		query, err = normalizeURIComponent(rest[queryAt+1:], uriQueryFragment)
		if err != nil {
			return "", err
		}
		rest = rest[:queryAt]
	}

	authority := ""
	hasAuthority := strings.HasPrefix(rest, "//")
	path := rest
	if hasAuthority {
		authorityEnd := strings.IndexByte(rest[2:], '/')
		if authorityEnd < 0 {
			authority = rest[2:]
			path = ""
		} else {
			authorityEnd += 2
			authority = rest[2:authorityEnd]
			path = rest[authorityEnd:]
		}
		var err error
		authority, err = normalizeURIAuthority(authority, scheme)
		if err != nil {
			return "", err
		}
	}

	path, err := normalizeURIComponent(path, uriPath)
	if err != nil {
		return "", err
	}
	path = removeURIDotSegments(path)

	var result strings.Builder
	result.Grow(len(value))
	result.WriteString(scheme)
	result.WriteByte(':')
	if hasAuthority {
		result.WriteString("//")
		result.WriteString(authority)
	}
	result.WriteString(path)
	if hasQuery {
		result.WriteByte('?')
		result.WriteString(query)
	}
	return result.String(), nil
}

type uriComponent int

const (
	uriPath uriComponent = iota
	uriQueryFragment
	uriUserInfo
	uriRegName
)

func normalizeURIComponent(value string, component uriComponent) (string, error) {
	var result strings.Builder
	result.Grow(len(value))
	for index := 0; index < len(value); index++ {
		character := value[index]
		if character == '%' {
			if index+2 >= len(value) || !isHex(value[index+1]) || !isHex(value[index+2]) {
				return "", hashV1Error(HashV1ErrorNormalization)
			}
			decoded := fromHex(value[index+1])<<4 | fromHex(value[index+2])
			if isURIUnreserved(decoded) {
				result.WriteByte(decoded)
			} else {
				result.WriteByte('%')
				result.WriteByte(toUpperHex(decoded >> 4))
				result.WriteByte(toUpperHex(decoded & 0x0f))
			}
			index += 2
			continue
		}
		if !isAllowedURICharacter(character, component) {
			return "", hashV1Error(HashV1ErrorNormalization)
		}
		result.WriteByte(character)
	}
	return result.String(), nil
}

func normalizeURIAuthority(authority, scheme string) (string, error) {
	userinfo := ""
	hostport := authority
	if at := strings.LastIndexByte(authority, '@'); at >= 0 {
		var err error
		userinfo, err = normalizeURIComponent(authority[:at], uriUserInfo)
		if err != nil {
			return "", err
		}
		userinfo += "@"
		hostport = authority[at+1:]
	}

	host := ""
	port := ""
	explicitPort := false
	bracketed := false
	if strings.HasPrefix(hostport, "[") {
		closeAt := strings.IndexByte(hostport, ']')
		if closeAt < 0 {
			return "", hashV1Error(HashV1ErrorNormalization)
		}
		bracketed = true
		host = hostport[1:closeAt]
		remainder := hostport[closeAt+1:]
		if remainder != "" {
			if !strings.HasPrefix(remainder, ":") {
				return "", hashV1Error(HashV1ErrorNormalization)
			}
			explicitPort = true
			port = remainder[1:]
		}
		if !isValidIPLiteral(host) {
			return "", hashV1Error(HashV1ErrorNormalization)
		}
		host = normalizeURIHostCase(host)
	} else {
		if strings.Count(hostport, ":") > 1 {
			return "", hashV1Error(HashV1ErrorNormalization)
		}
		if colonAt := strings.LastIndexByte(hostport, ':'); colonAt >= 0 {
			host = hostport[:colonAt]
			explicitPort = true
			port = hostport[colonAt+1:]
		} else {
			host = hostport
		}
		var err error
		host, err = normalizeURIComponent(host, uriRegName)
		if err != nil {
			return "", err
		}
		host = normalizeURIHostCase(host)
	}
	if host == "" || (explicitPort && port == "") {
		return "", hashV1Error(HashV1ErrorNormalization)
	}

	if port != "" {
		for index := range len(port) {
			if port[index] < '0' || port[index] > '9' {
				return "", hashV1Error(HashV1ErrorNormalization)
			}
		}
		if isDefaultPort(scheme, port) {
			port = ""
		}
	}

	if bracketed {
		host = "[" + host + "]"
	}
	if port != "" {
		host += ":" + port
	}
	return userinfo + host, nil
}

func removeURIDotSegments(path string) string {
	input := path
	output := make([]byte, 0, len(path))
	slashPositions := make([]int, 0, strings.Count(path, "/"))
	for input != "" {
		switch {
		case strings.HasPrefix(input, "../"):
			input = input[3:]
		case strings.HasPrefix(input, "./"):
			input = input[2:]
		case strings.HasPrefix(input, "/./"):
			input = input[2:]
		case input == "/.":
			input = "/"
		case strings.HasPrefix(input, "/../"):
			input = input[3:]
			output, slashPositions = removeLastURISegment(output, slashPositions)
		case input == "/..":
			input = "/"
			output, slashPositions = removeLastURISegment(output, slashPositions)
		case input == "." || input == "..":
			input = ""
		default:
			length := firstURISegmentLength(input)
			if input[0] == '/' {
				slashPositions = append(slashPositions, len(output))
			}
			output = append(output, input[:length]...)
			input = input[length:]
		}
	}
	return string(output)
}

func firstURISegmentLength(input string) int {
	start := 0
	if strings.HasPrefix(input, "/") {
		start = 1
	}
	if slashAt := strings.IndexByte(input[start:], '/'); slashAt >= 0 {
		return start + slashAt
	}
	return len(input)
}

func removeLastURISegment(output []byte, slashPositions []int) ([]byte, []int) {
	if len(slashPositions) > 0 {
		last := len(slashPositions) - 1
		return output[:slashPositions[last]], slashPositions[:last]
	}
	return output[:0], slashPositions
}

func isASCIIURI(value string) bool {
	for index := range len(value) {
		if value[index] < 0x21 || value[index] > 0x7e || value[index] == '\\' {
			return false
		}
	}
	return true
}

func isAllowedURICharacter(character byte, component uriComponent) bool {
	if isURIUnreserved(character) || strings.ContainsRune("!$&'()*+,;=", rune(character)) {
		return true
	}
	switch component {
	case uriPath:
		return character == '/' || character == ':' || character == '@'
	case uriQueryFragment:
		return character == '/' || character == '?' || character == ':' || character == '@'
	case uriUserInfo:
		return character == ':'
	case uriRegName:
		return false
	default:
		return false
	}
}

func isValidIPLiteral(value string) bool {
	if value == "" || strings.Contains(value, "%") {
		return false
	}
	if value[0] == 'v' || value[0] == 'V' {
		dotAt := strings.IndexByte(value, '.')
		if dotAt < 2 || dotAt == len(value)-1 {
			return false
		}
		for index := 1; index < dotAt; index++ {
			if !isHex(value[index]) {
				return false
			}
		}
		for index := dotAt + 1; index < len(value); index++ {
			character := value[index]
			if !isURIUnreserved(character) && !strings.ContainsRune("!$&'()*+,;=:", rune(character)) {
				return false
			}
		}
		return true
	}
	address, err := netip.ParseAddr(value)
	return err == nil && address.Is6()
}

func isDefaultPort(scheme, port string) bool {
	canonicalPort := strings.TrimLeft(port, "0")
	if canonicalPort == "" {
		canonicalPort = "0"
	}
	switch scheme {
	case "http":
		return canonicalPort == "80"
	case "https":
		return canonicalPort == "443"
	default:
		return false
	}
}

func normalizeURIHostCase(host string) string {
	var result strings.Builder
	result.Grow(len(host))
	for index := 0; index < len(host); index++ {
		character := host[index]
		if character == '%' && index+2 < len(host) && isHex(host[index+1]) && isHex(host[index+2]) {
			decoded := fromHex(host[index+1])<<4 | fromHex(host[index+2])
			if isURIUnreserved(decoded) {
				if decoded >= 'A' && decoded <= 'Z' {
					decoded += 'a' - 'A'
				}
				result.WriteByte(decoded)
			} else {
				result.WriteByte('%')
				result.WriteByte(toUpperHex(decoded >> 4))
				result.WriteByte(toUpperHex(decoded & 0x0f))
			}
			index += 2
			continue
		}
		if character >= 'A' && character <= 'Z' {
			character += 'a' - 'A'
		}
		result.WriteByte(character)
	}
	return result.String()
}

func isURIUnreserved(character byte) bool {
	return isASCIILetter(character) || (character >= '0' && character <= '9') ||
		character == '-' || character == '.' || character == '_' || character == '~'
}

func isASCIILetter(character byte) bool {
	return (character >= 'a' && character <= 'z') || (character >= 'A' && character <= 'Z')
}

func isHex(character byte) bool {
	return (character >= '0' && character <= '9') ||
		(character >= 'a' && character <= 'f') ||
		(character >= 'A' && character <= 'F')
}

func fromHex(character byte) byte {
	if character >= '0' && character <= '9' {
		return character - '0'
	}
	if character >= 'a' && character <= 'f' {
		return character - 'a' + 10
	}
	return character - 'A' + 10
}

func toUpperHex(value byte) byte {
	if value < 10 {
		return '0' + value
	}
	return 'A' + value - 10
}

File: defence_claw/defenseclaw/cli/defenseclaw/observability/hash_v1.py
"""Cross-language ``hash-v1`` redaction correlation primitive.

The public function returns only a non-reversible token.  This module never
loads keys from process state and its typed failures contain no caller values,
key identifiers, or key material.
"""

from __future__ import annotations

import hashlib
import hmac
import ipaddress
import re
import unicodedata
from dataclasses import dataclass
from enum import Enum

from defenseclaw.observability.unicode13 import _is_unicode13_repertoire

_DOMAIN = b"defenseclaw-redaction-hash-v1"
_KEY_BYTES = 32
_FIELD_CLASSES = frozenset(
    {
        "metadata",
        "identifier",
        "content",
        "reason",
        "evidence",
        "error",
        "path",
        "credential",
    }
)
_HIERARCHICAL_URI = re.compile(r"^[A-Za-z][A-Za-z0-9+.-]*://")
_SUB_DELIMS = frozenset("!$&'()*+,;=")
_HEX_DIGITS = frozenset("0123456789abcdefABCDEF")


class HashV1ErrorCode(str, Enum):
    """Value-free failure codes shared with the Go implementation."""

    INVALID_UTF8 = "invalid_utf8"
    INVALID_KEY = "invalid_key"
    UNSUPPORTED_CLASS = "unsupported_class"
    UNICODE_REPERTOIRE = "unicode_repertoire"
    NORMALIZATION_FAILED = "normalization_failed"


@dataclass(eq=False)
class HashV1Error(ValueError):
    """A typed, value-safe hash-v1 failure."""

    code: HashV1ErrorCode

    def __str__(self) -> str:
        return f"redaction hash-v1 failed: {self.code.value}"


def hash_v1(
    value: str | bytes,
    field_class: str,
    key: bytes,
) -> str:
    """Return the v1 keyed token for *value* without exposing normalization."""

    decoded = _decode_utf8(value)
    if not isinstance(key, bytes) or len(key) != _KEY_BYTES:
        raise HashV1Error(HashV1ErrorCode.INVALID_KEY)
    if field_class not in _FIELD_CLASSES:
        raise HashV1Error(HashV1ErrorCode.UNSUPPORTED_CLASS)

    normalized = _normalize_hash_v1_value(decoded, field_class)
    message = b"\x00".join((_DOMAIN, field_class.encode("ascii"), normalized.encode("utf-8")))
    digest = hmac.new(key, message, hashlib.sha256).hexdigest()
    key_id = hashlib.sha256(key).hexdigest()[:12]
    original_length = len(decoded.encode("utf-8"))
    return f"<hashed class={field_class} v=1 key={key_id} len={original_length} hmac={digest}>"


def _decode_utf8(value: str | bytes) -> str:
    if isinstance(value, bytes):
        try:
            return value.decode("utf-8", errors="strict")
        except UnicodeDecodeError:
            pass
        raise HashV1Error(HashV1ErrorCode.INVALID_UTF8)
    if isinstance(value, str):
        try:
            value.encode("utf-8", errors="strict")
        except UnicodeEncodeError:
            pass
        else:
            return value
        raise HashV1Error(HashV1ErrorCode.INVALID_UTF8)
    raise HashV1Error(HashV1ErrorCode.INVALID_UTF8)


def _normalize_hash_v1_value(value: str, field_class: str) -> str:
    if field_class not in _FIELD_CLASSES:
        raise HashV1Error(HashV1ErrorCode.UNSUPPORTED_CLASS)
    if not _is_unicode13_repertoire(value):
        raise HashV1Error(HashV1ErrorCode.UNICODE_REPERTOIRE)
    try:
        value.encode("utf-8", errors="strict")
    except UnicodeEncodeError:
        pass
    else:
        value = unicodedata.normalize("NFC", value)
        if field_class != "path":
            return value
        if _is_windows_drive_path(value):
            return _normalize_lexical_path(value)
        if _HIERARCHICAL_URI.match(value) is not None:
            return _normalize_absolute_uri(value)
        return _normalize_lexical_path(value)
    # Raise after leaving the handler so the safe error retains no value-bearing
    # Unicode exception as its cause or context.
    raise HashV1Error(HashV1ErrorCode.INVALID_UTF8)


def _is_windows_drive_path(value: str) -> bool:
    return len(value) >= 2 and _is_ascii_letter(value[0]) and value[1] == ":"


def _normalize_lexical_path(value: str) -> str:
    value = value.replace("\\", "/")
    prefix = ""
    absolute = False
    unc_root = False
    rest = value
    if rest.startswith("//"):
        prefix = "//"
        rest = rest[2:].lstrip("/")
        parts = [part for part in rest.split("/") if part]
        if len(parts) >= 2 and all(part not in (".", "..") for part in parts[:2]):
            prefix += f"{parts[0]}/{parts[1]}"
            rest = "/".join(parts[2:])
            absolute = True
            unc_root = True
    elif rest.startswith("/"):
        prefix = "/"
        absolute = True
        rest = rest[1:].lstrip("/")
    elif len(rest) >= 2 and _is_ascii_letter(rest[0]) and rest[1] == ":":
        prefix = rest[0].lower() + ":"
        rest = rest[2:]
        if rest.startswith("/"):
            prefix += "/"
            absolute = True
            rest = rest.lstrip("/")

    segments: list[str] = []
    for segment in rest.split("/"):
        if segment in ("", "."):
            continue
        if segment == "..":
            if segments and segments[-1] != "..":
                segments.pop()
            elif not absolute:
                segments.append(segment)
        else:
            segments.append(segment)

    joined = "/".join(segments)
    if not prefix:
        return joined
    if not joined:
        return prefix
    if prefix.endswith("/"):
        return prefix + joined
    if unc_root:
        return prefix + "/" + joined
    return prefix + joined


def _normalize_absolute_uri(value: str) -> str:
    if not _is_ascii_uri(value):
        raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
    colon = value.find(":")
    if colon <= 0:
        raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
    scheme = value[:colon].lower()
    rest = value[colon + 1 :]
    if not rest.startswith("//"):
        raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)

    fragment_at = rest.find("#")
    if fragment_at >= 0:
        _normalize_uri_component(rest[fragment_at + 1 :], "query_fragment")
        rest = rest[:fragment_at]

    query = ""
    has_query = False
    query_at = rest.find("?")
    if query_at >= 0:
        has_query = True
        query = _normalize_uri_component(rest[query_at + 1 :], "query_fragment")
        rest = rest[:query_at]

    authority = ""
    has_authority = rest.startswith("//")
    path = rest
    if has_authority:
        authority_end = rest.find("/", 2)
        if authority_end < 0:
            authority = rest[2:]
            path = ""
        else:
            authority = rest[2:authority_end]
            path = rest[authority_end:]
        authority = _normalize_uri_authority(authority, scheme)

    path = _normalize_uri_component(path, "path")
    path = _remove_uri_dot_segments(path)
    result = f"{scheme}:"
    if has_authority:
        result += f"//{authority}"
    result += path
    if has_query:
        result += f"?{query}"
    return result


def _normalize_uri_component(value: str, component: str) -> str:
    result: list[str] = []
    index = 0
    while index < len(value):
        character = value[index]
        if character == "%":
            if index + 2 >= len(value) or value[index + 1] not in _HEX_DIGITS or value[index + 2] not in _HEX_DIGITS:
                raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
            decoded = chr(int(value[index + 1 : index + 3], 16))
            if _is_uri_unreserved(decoded):
                result.append(decoded)
            else:
                result.append("%" + value[index + 1 : index + 3].upper())
            index += 3
            continue
        if not _is_allowed_uri_character(character, component):
            raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
        result.append(character)
        index += 1
    return "".join(result)


def _normalize_uri_authority(authority: str, scheme: str) -> str:
    userinfo = ""
    hostport = authority
    at = authority.rfind("@")
    if at >= 0:
        userinfo = _normalize_uri_component(authority[:at], "userinfo") + "@"
        hostport = authority[at + 1 :]

    bracketed = hostport.startswith("[")
    port = ""
    explicit_port = False
    if bracketed:
        close_at = hostport.find("]")
        if close_at < 0:
            raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
        host = hostport[1:close_at]
        remainder = hostport[close_at + 1 :]
        if remainder:
            if not remainder.startswith(":"):
                raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
            explicit_port = True
            port = remainder[1:]
        if not _is_valid_ip_literal(host):
            raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
        host = _normalize_uri_host_case(host)
    else:
        if hostport.count(":") > 1:
            raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
        if ":" in hostport:
            host, port = hostport.rsplit(":", 1)
            explicit_port = True
        else:
            host = hostport
        host = _normalize_uri_host_case(_normalize_uri_component(host, "reg_name"))

    if not host or (explicit_port and not port):
        raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)

    if port:
        if not port.isascii() or not port.isdecimal():
            raise HashV1Error(HashV1ErrorCode.NORMALIZATION_FAILED)
        if _is_default_port(scheme, port):
            port = ""

    if bracketed:
        host = f"[{host}]"
    if port:
        host += f":{port}"
    return userinfo + host


def _remove_uri_dot_segments(path: str) -> str:
    input_buffer = path
    output = bytearray()
    slash_positions: list[int] = []
    while input_buffer:
        if input_buffer.startswith("../"):
            input_buffer = input_buffer[3:]
        elif input_buffer.startswith("./"):
            input_buffer = input_buffer[2:]
        elif input_buffer.startswith("/./"):
            input_buffer = input_buffer[2:]
        elif input_buffer == "/.":
            input_buffer = "/"
        elif input_buffer.startswith("/../"):
            input_buffer = input_buffer[3:]
            _remove_last_uri_segment(output, slash_positions)
        elif input_buffer == "/..":
            input_buffer = "/"
            _remove_last_uri_segment(output, slash_positions)
        elif input_buffer in (".", ".."):
            input_buffer = ""
        else:
            length = _first_uri_segment_length(input_buffer)
            if input_buffer[0] == "/":
                slash_positions.append(len(output))
            output.extend(input_buffer[:length].encode("ascii"))
            input_buffer = input_buffer[length:]
    return output.decode("ascii")


def _first_uri_segment_length(value: str) -> int:
    start = 1 if value.startswith("/") else 0
    slash_at = value.find("/", start)
    return len(value) if slash_at < 0 else slash_at


def _remove_last_uri_segment(value: bytearray, slash_positions: list[int]) -> None:
    if slash_positions:
        del value[slash_positions.pop() :]
    else:
        value.clear()


def _is_ascii_uri(value: str) -> bool:
    return all("!" <= character <= "~" and character != "\\" for character in value)


def _is_allowed_uri_character(character: str, component: str) -> bool:
    if _is_uri_unreserved(character) or character in _SUB_DELIMS:
        return True
    if component == "path":
        return character in "/:@"
    if component == "query_fragment":
        return character in "/?:@"
    if component == "userinfo":
        return character == ":"
    return False


def _is_valid_ip_literal(value: str) -> bool:
    if not value or "%" in value:
        return False
    if value[0] in "vV":
        dot_at = value.find(".")
        if dot_at < 2 or dot_at == len(value) - 1:
            return False
        if any(character not in _HEX_DIGITS for character in value[1:dot_at]):
            return False
        return all(
            _is_uri_unreserved(character) or character in _SUB_DELIMS or character == ":"
            for character in value[dot_at + 1 :]
        )
    try:
        return ipaddress.ip_address(value).version == 6
    except ValueError:
        return False


def _is_default_port(scheme: str, port: str) -> bool:
    canonical_port = port.lstrip("0") or "0"
    return (scheme, canonical_port) in {
        ("http", "80"),
        ("https", "443"),
    }


def _normalize_uri_host_case(host: str) -> str:
    result: list[str] = []
    index = 0
    while index < len(host):
        character = host[index]
        if (
            character == "%"
            and index + 2 < len(host)
            and host[index + 1] in _HEX_DIGITS
            and host[index + 2] in _HEX_DIGITS
        ):
            decoded = chr(int(host[index + 1 : index + 3], 16))
            if _is_uri_unreserved(decoded):
                result.append(decoded.lower())
            else:
                result.append("%" + host[index + 1 : index + 3].upper())
            index += 3
            continue
        result.append(character.lower())
        index += 1
    return "".join(result)


def _is_uri_unreserved(character: str) -> bool:
    return _is_ascii_letter(character) or character.isascii() and character.isdecimal() or character in "-._~"


def _is_ascii_letter(character: str) -> bool:
    return "a" <= character <= "z" or "A" <= character <= "Z"
